### DISCUSSION TITLE
Welcome to datamesh-architecture.com Discussions!

### DISCUSSION THREAD

**Discussion by simonharrer** [datamesh-architecture/datamesh-architecture.com#3] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by kmerroll** [comment]
Interesting article giving an overview of the data mesh architecture, but as a information architect I feel there are some significant gaps in applying a domain-driven design methodology to the data mesh architecture described in the article. These are very deep topics so I'll need to summarize the overall feedback:

1. Firstly, domain-driven design (DDD) is a attractive architecture pattern to address the centralized management approach used in many organizations. However I feel the article needs to describe more about how DDD is implemented. Specifically, domains are categories or groupings of concepts, resources, functions, organizations, capabilities, etc. The article references "Domains" but not "Domains of what?" question. Many organizations struggle with DDD because of the ambiguity about how to organize their domains and usually default to functional alignment between development teams which requires a top-down approach to create the taxonomy or common namespace for these functional areas.
2. The event-driven architecture approach of having the repositories (System of Record, System of Reference) emit lifecycle events (create, update, delete) to subscribers requires the event consumers to implement event orchestration logic to trigger data synchronization or other "Thick" event requests via API or other brokers. This great increases the complexity of the data distribution pipelines, development overhead, and operational support. The other approaches for data synchronization such as continuous queries, change-data-capture, or even the dreaded truncate and load all have their issues, but using event-driven architecture approach that works well in tightly bound interface (GUI) frameworks seems like a "hammer" looking for "nails" approach.
3. I didn't see any mention of Streaming data architecture using distributed journal (Kafka) or complex-event-processing (Flink, etc.) to address the "Data Mesh". Arguably the streaming or "Kappa" architecture pattern for distributed data platforms would be as good if not far superior to the domain-driven / event-driven approach proposed in the article.
4. I liked that you included Data Governance in the overall Data Mesh platform, but think there needs to be more detail than appending "Agile" to the name. Specifically, distributed data framework have an even higher need for data governance related to the core governance areas of Master Data Management, Data Quality Management, and Change Management. Also, the policies and compliance controls referenced in the article are commendable, but are missing core details in data security classification, data privacy regulations (GDPR, CCPA), and role-based access permissions.

Okay, lots more to add related to the data catalog, semantic search, and data workspace environments, but these are long topics. 

Kudos on putting the time into an interesting article. I think you're close to describing the current modern data architecture patterns and hoping this constructive feedback is helpful.

-Kimi

**Comment by aliansari** [comment]
Very good and practical article. Thank you for the effort.

I have a question. 

We are trying to adopt data mesh in my organization. Can you please guide on what is the best practice when it comes to data product exchange? Can consumer pull the product from producer? or produce can push the product to consumer?

In our case we are expecting the data producer to create the data product for the consumer on a daily basis.

Happy to provide more information if required.

Thanks

**Reply by jochenchrist** [reply]
Hi aliansari,
great, that you liked the content :)

My answer would be:
Producers publish data in their managed space (like files in a S3 Bucket) and they give appropriate permissions to consuming teams. Consumers pull data. Actually, depending on the data platforms, consumers don't actually need to copy the data, but just link the data and join or transform it to build new data products (in their space)

The microservice equivalent would be GET API endpoints. Consumers can pull, if they need data.

**Comment by aliansari** [comment]
Thanks for the quick reply.

We have a policy issue where we need to persist the data in the domain. I guess that's why we need to pull/push data into the consumer domain, store it in S3 and then perform further transformations. This particular consumer will create couple of data products for other domains to consume.

I'll go through the article few more times to digest it fully :)

**Reply by jochenchrist** [reply]
If you have a policy that the consumer needs to persist data, then it is totally fine to copy the data. Basically, the consuming team becomes an owner of their data copy, with full lifecycle responsibilities.

**Reply by aliansari** [reply]
yes, that's pretty much the idea.

**Comment by 0xxx** [comment]
Great self-healing data article and nice data-product practicer keener than ever!
I am on tailwinds towards exploratory data analysis with an approach on data mesh and machine learning.

**Comment by AlexanderVanHelm** [comment]
Can we reuse the beautiful drawings to explain data mesh in a webinar?

**Reply by jochenchrist** [reply]
@AlexanderVanHelm Feel free to use the images in a webinar. Thanks for asking.

**Comment by shaikidris** [comment]
This is an excellent articulation of decentralized data-teams. While a lot of data innovations have hapend across large scale B2C consumer facing companies (ex: uber, airbn, linkedin) we do not have much from enterprise SaaS companies.

I am in the process of dataproduct transformation at my org which is an enterprise SaaS for banks. We have multiple domains like Credit account processing, Loans, Transaction processing etc. Modeling the domains ( and respective teams) using data-mesh architecture simplifies so many things. However, this is not a technology problem rather cultural/mindset problem for devs :)
